File: LeetCode/src/binarysearch/SqrtX.java
package binarysearch;

public class SqrtX {
    public int mySqrt(int x) {
        int left = 0, right = x;
        int result = 0;

        while (left <= right) {
            int mid = left + (right - left) / 2;
            if ((long) mid * mid <= x) {
                result = mid;
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }

        return result;
    }

    public static void main(String[] args) {
        SqrtX solution = new SqrtX();
        int x1 = 4;
        int x2 = 8;
        System.out.println(solution.mySqrt(x1));// Output: 2
        System.out.println(solution.mySqrt(x2));// Output: 2
    }
}

File: LeetCode_Python/binarysearch/SqrtX.py
class SqrtX:
    def mySqrt(self, x: int) -> int:
        if x < 2:
            return x

        left, right = 2, x // 2

        while left <= right:
            mid = left + (right - left) // 2
            num = mid * mid
            if num < x:
                left = mid + 1
            elif num > x:
                right = mid - 1
            else:
                return mid

        # When the loop ends, left is greater than right.
        # The largest integer whose square is less than or equal to x is right.
        # This is because left will be the first number whose square is greater than x.
        return right


# Time Complexity: O(log n), where n is the input number x. The binary search reduces the search space by half in each iteration.
# Space Complexity: O(1), as we are using a constant amount of space for variables.

if __name__ == "__main__":
    solution = SqrtX()

    print("Expected Output: 2")
    print("Actual Output:", solution.mySqrt(8))

    print("Expected Output: 4")
    print("Actual Output:", solution.mySqrt(16))
